Objects/listobject.c

```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
         reverse_slice(s->values, &s->values[n]);
 }
 
+/* Here we define custom comparison functions to optimize for the cases one commonly 
+ * encounters in practice: homogeneous lists, often of one of the basic types. */
+
+/* This struct holds the comparison function and helper functions 
+ * selected in the pre-sort check. */
+
+/* These are the special case compare functions.
+ * ms->key_compare will always point to one of these: */
+
+/* Heterogeneous compare: default, always safe to fall back on. */
+static int
+safe_object_compare(PyObject* v, PyObject* w, MergeState* ms)
+{
+  /* No assumptions necessary! */
+    return PyObject_RichCompareBool(v, w, Py_LT);
+}
+
+/* Homogeneous compare: safe for any two compareable objects of the same type.
+ * (ms->key_richcompare is set to ob_type->tp_richcompare in the
+ *  pre-sort check.)
+ */
+static int
+unsafe_object_compare(PyObject* v, PyObject* w, MergeState* ms)
+{
+  /* Modified from Objects/object.c:PyObject_RichCompareBool, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
```

Comment:
Yes, you are correct. `ppperry` gave an explicit counterexample to this. It's OK, though, because I added the following:
```
    if (v->ob_type->tp_richcompare != ms->key_richcompare)
        return PyObject_RichCompareBool(v, w, Py_LT);
```
However, I should get rid of that assertion.